src/LinearSolverTypes.f90

```diff
@@ -1597,228 +1629,237 @@ SUBROUTINE solveBiCGSTAB(solver)
-        v(:,it+1)=w/h
-      ELSE
-        v(:,it+1)=0.0_SRK*w
-      ENDIF
-      !Set up next Given's rotation
-      IF(t >= 0.0_SRK) THEN
-        temp=SQRT(t*t+h*h)
-      ELSE
-        temp=-SQRT(t*t+h*h)
-      ENDIF
-      c(it)=t/temp
-      s(it)=h/temp
-      R(it,it)=temp
-      g(it)=c(it)*phibar
-      phibar=-s(it)*phibar
-#ifdef FUTILITY_DEBUG_MSG
-      WRITE(668,*) '         GMRES-NOPC',it,ABS(phibar)
-#endif
-      IF(ABS(phibar) <= tol) EXIT
-    ENDDO
+ENDSUBROUTINE solveGMRES
 
-    y(1:it)=g(1:it)
-    CALL BLAS_matvec('U','N','N',R(1:it,1:it),y(1:it))
-
-    CALL BLAS_matvec(v(:,1:it),y(1:it),0.0_SRK,u%b)
-    CALL BLAS_axpy(u,solver%x)
-    CALL solver%getResidual(u)
-    CALL LNorm(u%b,2,beta)
-    IF(it == m+1) it=m
-    solver%iters=it
-
-    DEALLOCATE(v)
-    DEALLOCATE(R)
-    DEALLOCATE(w)
-    DEALLOCATE(c)
-    DEALLOCATE(s)
-    DEALLOCATE(g)
-    DEALLOCATE(y)
-  ENDIF
-  solver%info=0
-  CALL u%clear()
-ENDSUBROUTINE solveGMRES_nopc
 !
 !-------------------------------------------------------------------------------
-!> @brief Solves the Iterative Linear System using the GMRES method with
-!>        left-preconditioning
+!> @brief GMRES core solver routines.
 !> @param solver The linear solver to act on
 !> @param PreCondType The preconditioner object to use on the system
 !>
 !> This subroutine solves the Iterative Linear System using the GMRES method
 !>
-SUBROUTINE solveGMRES_lpc(solver,PreCondType)
-  CLASS(LinearSolverType_Iterative),INTENT(INOUT) :: solver
-  CLASS(PreconditionerType),INTENT(INOUT) :: PrecondType
+SUBROUTINE solveGMRES_partial(thisLS,nIters,converged,thisPC)
+  CLASS(LinearSolverType_Iterative),INTENT(INOUT) :: thisLS
+  CLASS(PreconditionerType),INTENT(INOUT),OPTIONAL :: thisPC
+  INTEGER(SIK),INTENT(OUT) :: nIters
+  LOGICAL(SBK),INTENT(OUT) :: converged
+
+  ! Array of orthogonal basis vectors
+  CLASS(NativeVectorType),ALLOCATABLE :: V(:)
+  CLASS(NativeVectorType),ALLOCATABLE :: u,Vy ! Generic vector container
+  REAL(SRK),ALLOCATABLE :: R(:,:) ! Array of basis vector coeffs for sol.
+  REAL(SRK),ALLOCATABLE :: givens_sin(:),givens_cos(:),f(:),h(:)
+  TYPE(ParamType) :: vecPlist
+  REAL(SRK) :: norm_b,norm_r0,currResid
+  REAL(SRK) :: divTmp,t,temp,tol
+  INTEGER(SIK) :: krylovIdx,orthogIdx,nLocal,i,j,oi1,mpierr
+  INTEGER(SIK),ALLOCATABLE :: orthogReq(:)
+
+  CALL vecPlist%clear()
+  SELECT TYPE(x => thisLS%X)
+  TYPE IS(RealVectorType)
+    CALL vecPlist%add('VectorType -> n',thisLS%A%n)
+    ALLOCATE(RealVectorType :: u)
+    ALLOCATE(RealVectorType :: Vy)
+#ifdef HAVE_MPI
+  TYPE IS(NativeDistributedVectorType)
+    CALL vecPlist%add('VectorType -> n',thisLS%A%n)
+    CALL vecPlist%add('VectorType -> MPI_Comm_ID',thisLS%MPIparallelEnv%comm)
+    CALL vecPlist%add('VectorType -> chunkSize',x%chunkSize)
+    CALL vecPlist%add('VectorType -> nlocal',SIZE(x%b))
+    ALLOCATE(NativeDistributedVectorType :: u)
+    ALLOCATE(NativeDistributedVectorType :: Vy)
```

Comment:
Another possible `#else` for the serial portion?

Reply:
This would prevent one from running a solver in serial when using a parallel build; which seems like an unnecessary restriction to me. If you believe it's the better course of action, we can pursue it though.